tests/framework/crossplane.go

```diff
@@ -0,0 +1,218 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+// ExpectedNginxField contains an nginx directive key and value,
+// and the expected file, server, and location block that it should exist in.
+type ExpectedNginxField struct {
+	// Key is the directive name.
+	Key string
+	// Value is the value for the directive. Can be the full value or a substring.
+	Value string
+	// File is the file name that should contain the directive. Can be a full filename or a substring.
+	File string
+	// Location is the location name that the directive should exist in.
+	Location string
+	// Servers are the server names that the directive should exist in.
+	Servers []string
+	// ValueSubstringAllowed allows the expected value to be a substring of the real value.
+	// This makes it easier for cases when real values are complex file names or contain things we
+	// don't care about, and we just want to check if a substring exists.
+	ValueSubstringAllowed bool
+}
+
+// ValidateNginxFieldExists accepts the nginx config and the configuration for the expected field,
+// and returns whether or not that field exists where it should.
+func ValidateNginxFieldExists(conf *Payload, expFieldCfg ExpectedNginxField) bool {
+	for _, config := range conf.Config {
+		if !strings.Contains(config.File, expFieldCfg.File) {
+			continue
+		}
+
+		for _, directive := range config.Parsed {
+			if len(expFieldCfg.Servers) == 0 {
+				if expFieldCfg.fieldFound(directive) {
+					return true
+				}
+				continue
+			}
+
+			for _, serverName := range expFieldCfg.Servers {
+				if directive.Directive == "server" && getServerName(directive.Block) == serverName {
+					for _, serverDirective := range directive.Block {
+						if expFieldCfg.Location == "" && expFieldCfg.fieldFound(serverDirective) {
+							return true
+						} else if serverDirective.Directive == "location" &&
+							fieldExistsInLocation(serverDirective, expFieldCfg) {
+							return true
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func getServerName(serverBlock Directives) string {
+	for _, directive := range serverBlock {
+		if directive.Directive == "server_name" {
+			return directive.Args[0]
+		}
+	}
+
+	return ""
+}
+
+func (e ExpectedNginxField) fieldFound(directive *Directive) bool {
+	arg := strings.Join(directive.Args, " ")
+
+	valueMatch := arg == e.Value
+	if e.ValueSubstringAllowed {
+		valueMatch = strings.Contains(arg, e.Value)
+	}
+
+	return directive.Directive == e.Key && valueMatch
+}
+
+func fieldExistsInLocation(serverDirective *Directive, expFieldCfg ExpectedNginxField) bool {
```

Comment:
```suggestion
func fieldExistsInLocation(locationDirective *Directive, expFieldCfg ExpectedNginxField) bool {
```

`serverDirective` confused me because I thought we were passing in the `server` directive.